plugins/admin.py

```diff
@@ -0,0 +1,131 @@
+"""
+import asyncio
+from time import time
+from pyrogram import Client, filters
+from pyrogram.types import Message
+from pyrogram.errors import UserNotParticipant
+from config import ADMIN_CHATS
+
+USAGE_KICK = """**Usage**:
+*admin reply to a user message*
+`!ban [reason]`
+"""
+USAGE_BAN = """**Usage**:
+*admin reply to a user message*
+`!ban [reason]`
+"""
+USAGE_UNBAN = """**Usage**:
+*admin reply to a user message*
+`!unban [reason]`
+"""
+
+# should be more than 30 seconds
+# or less than 366 days
+DURATION_UNTIL_UNBAN = 60
+DELAY_DELETE = 15
+
+
+@Client.on_message(filters.command(["kick"], prefixes="!")
+                   & filters.chat(ADMIN_CHATS)
+                   & filters.incoming
+                   & ~filters.edited)
+async def kick_user(_, message: Message):
+    """!kick admin kick user"""
+    try:
+        if not (await _should_allow_admin_action(message)):
+            reply = await message.reply_text(USAGE_KICK)
+            await _delay_delete_messages((reply, message), DELAY_DELETE)
+            return
+    except UserNotParticipant:
+        await message.reply_text("Target user is not member of this chat")
+        return
+    reason = " ".join(message.command[1:])
+    u_proposer = message.from_user
+    u_target = message.reply_to_message.from_user
+    await message.chat.kick_member(u_target.id,
+                                   int(time() + DURATION_UNTIL_UNBAN))
+    await message.reply_text("{} kicked {}\n**Reason**: {}"
+                             .format(u_proposer.mention,
+                                     u_target.mention,
+                                     reason))
+    await message.delete()
+
+
+@Client.on_message(filters.command(["ban"], prefixes="!")
+                   & filters.chat(ADMIN_CHATS)
+                   & filters.incoming
+                   & ~filters.edited)
+async def ban_user(_, message: Message):
+    """!ban admin ban members"""
+    try:
+        if not (await _should_allow_admin_action(message)):
+            reply = await message.reply_text(USAGE_BAN)
+            await _delay_delete_messages((reply, message), DELAY_DELETE)
+            return
+    except UserNotParticipant:
+        await message.reply_text("Target user is not member of this chat")
+        return
+    reason = " ".join(message.command[1:])
+    u_proposer = message.from_user
+    u_target = message.reply_to_message.from_user
+    await message.chat.kick_member(u_target.id)
+    await message.reply_text("{} banned {}\n**Reason**: {}"
+                             .format(u_proposer.mention,
+                                     u_target.mention,
+                                     reason))
+    await message.delete()
+
+
+@Client.on_message(filters.command(["unban"], prefixes="!")
+                   & filters.chat(ADMIN_CHATS)
+                   & filters.incoming
+                   & ~filters.edited)
+async def unban_user(_, message: Message):
+    """!unban admin unban members"""
+    try:
```

Comment:
this chunk of code:

```
    try:
        if not (await _should_allow_admin_action(message)):
            reply = await message.reply_text(USAGE_BAN)
            await _delay_delete_messages((reply, message), DELAY_DELETE)
            return
    except UserNotParticipant:
        await message.reply_text("Target user is not member of this chat")
        return
```
duplicated **3 times**,  did you consider moving it to a method?

maybe method like validateAdmin(message: Message, userActionWarning: str) that returns boolean

Reply:
i think for these methods you're writing too much of the same code, what do you think of doing it like this?

```
@Client.on_message(filters.command(["kick"], prefixes="!")
                   & filters.chat(ADMIN_CHATS)
                   & filters.incoming
                   & ~filters.edited)
async def kick_user(_, message: Message):
    """!kick admin kick user"""
    kick_allowed = await _validate_user(message, USAGE_KICK)
    if kick_allowed:
        return
    await message.chat.kick_member(message.reply_to_message.from_user.id,
                                   int(time() + DURATION_UNTIL_UNBAN))
    await _reply_action_succeeded(message, "kicked")
    await message.delete()


@Client.on_message(filters.command(["ban"], prefixes="!")
                   & filters.chat(ADMIN_CHATS)
                   & filters.incoming
                   & ~filters.edited)
async def ban_user(_, message: Message):
    """!ban admin ban members"""
    ban_allowed = await _validate_user(message, USAGE_BAN)
    if ban_allowed:
        return
    await message.chat.kick_member(message.reply_to_message.from_user.id)
    await _reply_action_succeeded(message, "banned")
    await message.delete()


@Client.on_message(filters.command(["unban"], prefixes="!")
                   & filters.chat(ADMIN_CHATS)
                   & filters.incoming
                   & ~filters.edited)
async def unban_user(_, message: Message):
    """!unban admin unban members"""
    unban_allowed = await _validate_user(message, USAGE_UNBAN)
    if unban_allowed:
        return
    await message.chat.unban_member(message.reply_to_message.from_user.id)
    await _reply_action_succeeded(message, "unbanned")
    await message.delete()


async def _validate_user(message: Message, user_action_warning: str):
    try:
        if not (await _should_allow_admin_action(message)):
            reply = await message.reply_text(user_action_warning)
            await _delay_delete_messages((reply, message), DELAY_DELETE)
            return True
    except UserNotParticipant:
        await message.reply_text("Target user is not member of this chat")
        return True
    return False


async def _reply_action_succeeded(message: Message, performed_action: str):
    reason = " ".join(message.command[1:])
    u_proposer = message.from_user
    u_target = message.reply_to_message.from_user
    await message.reply_text("{} " + performed_action + " {}\n**Reason**: {}"
                             .format(u_proposer.mention,
                                     u_target.mention,
                                     reason))
```

**P.S.** i didn't test my code snippet, it's just an idea for you to which direction you can go

Reply:
`_reply_action_succeed` is good, for the rest part, I'll consider after add a custom filter.

I got a question, what if a spammer join the group, post ads and leave the group? it will match `UserNotParticipant`, but in this case the admin may want to ban him. but for `!kick` it should not allow the admin action.

would be a problem for those commands share such function, not sure if it worth a parameter for this.

Anyway, I'll think of that after add a custom filter to exclude such:
- messages comes from chat (could be anonymous admin or messages from the linked channel), I think it's a bad idea to let anonymous admin to ban user
- message replies to anonymous admin

Reply:
> but in this case the admin may want to ban him. but for !kick it should not allow the admin action. would be a problem for those commands share such function, not sure if it worth a parameter for this.

it's a good point, i agree

> I think it's a bad idea to let anonymous admin to ban user

i think you're right